[1.10] Docs for Cloudflare components (#3032)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-state-stores/setup-cloudflare-workerskv.md

|--------------------|:--------:|---------|---------|
| `kvNamespaceID` | Y | ID of the pre-created Workers KV namespace | `"123456789abcdef8b5588f3d134f74ac"`
| `workerName` | Y | Name of the Worker to connect to | `"mydaprkv"`
| `key` | Y | Ed25519 private key, PEM-encoded | *See example above*
| `cfAccountID` | Y/N | Cloudflare account ID. Required to have Dapr manage the worker. | `"456789abcdef8b5588f3d134f74ac"def`
| `cfAPIToken` | Y/N | API token for Cloudflare. Required to have Dapr manage the Worker. | `"secret-key"`
| `workerUrl` | Y/N | URL of the Worker. Required if the Worker has been pre-provisioned outside of Dapr. | `"https://mydaprkv.mydomain.workers.dev"`

> When you configure Dapr to create your Worker for you, you may need to set a longer value for the `initTimeout` property of the component, to allow enough time for the Worker script to be deployed. For example: `initTimeout: "120s"`

## Create a Workers KV namespace

To use this component, you must have a Workers KV namespace created in your Cloudflare account.

You can create a new Workers KV namespace in one of two ways:

- Using the [Cloudflare dashboard](https://dash.cloudflare.com/)  

>   Make note of the "ID" of the Workers KV namespace that you can see in the dashboard. This is a hex string (for example `123456789abcdef8b5588f3d134f74ac`)–not the name you used when you created it!

Reviewer: ```suggestion
  Make note of the "ID" of the Workers KV namespace that you can see in the dashboard. This is a hex string (for example `123456789abcdef8b5588f3d134f74ac`) – not the name you used when you created it!
```

Author: This is actually an em-dash (GitHub doesn't render it well, but you can see it in VS Code:
![image](https://user-images.githubusercontent.com/43508/210644031-dc7e8ce0-db4e-4e87-8636-47f7068a5761.png)

My understanding is that with em-dashes there should not be a space around them

- Using the [Wrangler CLI](https://developers.cloudflare.com/workers/wrangler/):
  
  ```sh
  # Authenticate if needed with `npx wrangler login` first
  wrangler kv:namespace create <NAME>
  ```

>   The output will contain the ID of the namespace, for example:

Reviewer: ```suggestion
  The output contains the ID of the namespace, for example:
```

Next, you can generate a new Ed25519 key pair with the step CLI using:

```sh
step crypto keypair \
  public.pem private.pem \
  --kty OKP --curve Ed25519 \
  --insecure --no-password
```

{{% /codetab %}}

{{< /tabs >}}

Regardless of how you generated your key pair, with the instructions above you'll have two files:

- `private.pem` contains the private part of the key: use the contents of this file for the **`key`** property of the component's metadata.
- `public.pem` contains the public part of the key, which you'll need only if you're deploying a Worker manually (as per the instructions in the previoius section).

{{% alert title="Warning" color="warning" %}}
Protect the private part of your key and treat it as a secret value!
{{% /alert %}}

## Additional notes

- Note that Cloudflare Workers KV doesn't guarantee strong data consistency. Although changes are visible immediately (usually) for requests made to the same Cloudflare datacenter, it can take a certain amount of time (usually up to one minute) for changes to be replicated across all Cloudflare regions.
- This state store supports TTLs with Dapr, but the minimum value for the TTL is 1 minute.

> ## Related links

Reviewer: Is it worth referencing any CloudFlare WorkerKV documentation here? It generally helps and is recommended.

Author: Done

You can create a new Workers KV namespace in one of two ways:

<!-- IGNORE_LINKS -->
- Using the [Cloudflare dashboard](https://dash.cloudflare.com/)  
  Make note of the "ID" of the Workers KV namespace that you can see in the dashboard. This is a hex string (for example `123456789abcdef8b5588f3d134f74ac`)–not the name you used when you created it!
- Using the [Wrangler CLI](https://developers.cloudflare.com/workers/wrangler/):
  
  ```sh
  # Authenticate if needed with `npx wrangler login` first
  wrangler kv:namespace create <NAME>
  ```

  The output contains the ID of the namespace, for example:

  ```text
  { binding = "<NAME>", id = "123456789abcdef8b5588f3d134f74ac" }
  ```
<!-- END_IGNORE -->

## Configuring the Worker

Because Cloudflare Workers KV namespaces can only be accessed by scripts running on Workers, Dapr needs to maintain a Worker to communicate with the Workers KV storage.

Dapr can manage the Worker for you automatically, or you can pre-provision a Worker yourself. Pre-provisioning the Worker is the only supported option when running on [workerd](https://github.com/cloudflare/workerd).

> > **Important:** Use a separate Worker for each Dapr component. Do not use the same Worker script for different Cloudflare Workers KV state store components, and do not use the same Worker script for different Cloudflare components in Dapr (e.g. the Workers KV state store and the Queues binding).

Reviewer 2: ```suggestion
{{% alert title="Important" color="warning" %}} 
Use a separate Worker for each Dapr component. Do not use the same Worker script for different Cloudflare Workers KV state store components, and do not use the same Worker script for different Cloudflare components in Dapr (e.g. the Workers KV state store and the Queues binding).
{{% /alert %}}
```

{{< tabs "Let Dapr manage the Worker" "Manually provision the Worker script" >}}

{{% codetab %}}
<!-- Let Dapr manage the Worker -->

If you want to let Dapr manage the Worker for you, you will need to provide these 3 metadata options:

<!-- IGNORE_LINKS -->

> - **`workerName`**: Name of the Worker script. This will be the first part of the URL of your Worker: for example, if the "workers.dev" domain configured for your Cloudflare account is `mydomain.workers.dev` and you set `workerName` to `mydaprkv`, the Worker that Dapr deploys will be available at `https://mydaprkv.mydomain.workers.dev`.

Reviewer 2: ```suggestion
- **`workerName`**: Name of the Worker script. This will be the first part of the URL of your Worker. For example, if the "workers.dev" domain configured for your Cloudflare account is `mydomain.workers.dev` and you set `workerName` to `mydaprkv`, the Worker that Dapr deploys will be available at `https://mydaprkv.mydomain.workers.dev`.
```

- **`cfAccountID`**: ID of your Cloudflare account. You can find this in your browser's URL bar after logging into the [Cloudflare dashboard](https://dash.cloudflare.com/), with the ID being the hex string right after `dash.cloudflare.com`. For example, if the URL is `https://dash.cloudflare.com/456789abcdef8b5588f3d134f74acdef`, the value for `cfAccountID` is `456789abcdef8b5588f3d134f74acdef`.

> - **`cfAPIToken`**: API token with permission to create and edit Workers and Workers KV namespaces. You can create it from the ["API Tokens" page](https://dash.cloudflare.com/profile/api-tokens) in the "My Profile" section in the Cloudflare dashboard. Click on "Create token", select the **"Edit Cloudflare Workers"** template, then follow the on-screen instructions to generate a new API token.

Reviewer 2: ```suggestion
- **`cfAPIToken`**: API token with permission to create and edit Workers and Workers KV namespaces. You can create it from the ["API Tokens" page](https://dash.cloudflare.com/profile/api-tokens) in the "My Profile" section in the Cloudflare dashboard: 
   1. Click on **"Create token"**.
   1. Select the **"Edit Cloudflare Workers"** template.
   1. Follow the on-screen instructions to generate a new API token.
```

- The bearer token is generated on each request and is valid for a brief period of time only (currently, one minute).

To let Dapr issue bearer tokens, and have your Worker validate them, you will need to generate a new Ed25519 key pair. Here are examples of generating the key pair using OpenSSL or the step CLI.

{{< tabs "Generate with OpenSSL" "Generate with the step CLI" >}}

{{% codetab %}}
<!-- Generate with OpenSSL -->

> Support for generating Ed25519 keys is available since OpenSSL 1.1.0, so the commands below will not work if you're using an older version of OpenSSL.

> Note for Mac users: on macOS, the "openssl" binary that is shipped by Apple is actually based on LibreSSL, which as of writing doesn't support Ed25519 keys. If you're using macOS, either use the step CLI, or install OpenSSL 3.0 from Homebrew using `brew install openssl@3` then replacing `openssl` in the commands below with `$(brew --prefix)/opt/openssl@3/bin/openssl`.

You can generate a new Ed25519 key pair with OpenSSL using:

```sh
openssl genpkey -algorithm ed25519 -out private.pem
openssl pkey -in private.pem -pubout -out public.pem

> # On Mac, using openssl@3 from Homebrew:

Reviewer 2: Same here. Ok to split into its own code block?

Author: Done

$(brew --prefix)/opt/openssl@3/bin/openssl genpkey -algorithm ed25519 -out private.pem
$(brew --prefix)/opt/openssl@3/bin/openssl pkey -in private.pem -pubout -out public.pem
```

{{% /codetab %}}

{{% codetab %}}
<!-- Generate with the step CLI -->

If you don't have the step CLI already, install it following the [official instructions](https://smallstep.com/docs/step-cli/installation).

Next, you can generate a new Ed25519 key pair with the step CLI using:

```sh
step crypto keypair \
  public.pem private.pem \
  --kty OKP --curve Ed25519 \
  --insecure --no-password
```

{{% /codetab %}}

{{< /tabs >}}

Regardless of how you generated your key pair, with the instructions above you'll have two files:

> - `private.pem` contains the private part of the key: use the contents of this file for the **`key`** property of the component's metadata.

Reviewer 2: ```suggestion
- `private.pem` contains the private part of the key; use the contents of this file for the **`key`** property of the component's metadata.
```
